File: app/src/main/java/com/coinninja/coinkeeper/service/runner/FullSyncWalletRunner.java
package com.coinninja.coinkeeper.service.runner;

import com.coinninja.coinkeeper.cn.account.RemoteAddressCache;
import com.coinninja.coinkeeper.cn.service.runner.AccountDeverificationServiceRunner;
import com.coinninja.coinkeeper.cn.wallet.CNWalletManager;
import com.coinninja.coinkeeper.model.helpers.WalletHelper;
import com.coinninja.coinkeeper.receiver.WalletSyncCompletedReceiver;
import com.coinninja.coinkeeper.util.Intents;
import com.coinninja.coinkeeper.util.android.LocalBroadCastUtil;

import org.greenrobot.greendao.DaoException;

import javax.inject.Inject;

public class FullSyncWalletRunner implements Runnable {
    private final CNWalletManager cnWalletManager;
    private final AccountDeverificationServiceRunner accountDeverificationServiceRunner;
    private final WalletRegistraionRunner walletRegistraionRunner;
    private final CurrentBTCStateRunner currentBTCStateRunner;
    private final SyncRunnable syncRunnable;
    private final TransactionConfirmationUpdateRunner transactionConfirmationUpdateRunner;
    private final FailedBroadcastCleaner failedBroadcastCleaner;
    private final SyncIncomingInvitesRunner syncIncomingInvitesRunner;
    private final FulfillSentInvitesRunner fulfillSentInvitesRunner;
    private final ReceivedInvitesStatusRunner receivedInvitesStatusRunner;
    private final NegativeBalanceRunner negativeBalanceRunner;
    private final WalletHelper walletHelper;
    private final LocalBroadCastUtil localBroadCastUtil;
    private final RemoteAddressCache remoteAddressCache;

    @Inject
    FullSyncWalletRunner(CNWalletManager cnWalletManager, AccountDeverificationServiceRunner accountDeverificationServiceRunner,
                         WalletRegistraionRunner walletRegistraionRunner, CurrentBTCStateRunner currentBTCStateRunner,
                         SyncRunnable syncRunnable, TransactionConfirmationUpdateRunner transactionConfirmationUpdateRunner,
                         FailedBroadcastCleaner failedBroadcastCleaner, SyncIncomingInvitesRunner syncIncomingInvitesRunner,
                         FulfillSentInvitesRunner fulfillSentInvitesRunner, ReceivedInvitesStatusRunner receivedInvitesStatusRunner,
                         NegativeBalanceRunner negativeBalanceRunner,
                         WalletHelper walletHelper, LocalBroadCastUtil localBroadCastUtil, RemoteAddressCache remoteAddressCache
    ) {

        this.cnWalletManager = cnWalletManager;
        this.accountDeverificationServiceRunner = accountDeverificationServiceRunner;
        this.walletRegistraionRunner = walletRegistraionRunner;
        this.currentBTCStateRunner = currentBTCStateRunner;
        this.syncRunnable = syncRunnable;
        this.transactionConfirmationUpdateRunner = transactionConfirmationUpdateRunner;
        this.failedBroadcastCleaner = failedBroadcastCleaner;
        this.syncIncomingInvitesRunner = syncIncomingInvitesRunner;
        this.fulfillSentInvitesRunner = fulfillSentInvitesRunner;
        this.receivedInvitesStatusRunner = receivedInvitesStatusRunner;
        this.negativeBalanceRunner = negativeBalanceRunner;
        this.walletHelper = walletHelper;
        this.localBroadCastUtil = localBroadCastUtil;
        this.remoteAddressCache = remoteAddressCache;
    }

    @Override
    public void run() {
        if (!cnWalletManager.hasWallet())
            return;

        try {
            syncTransactions();
            if (walletHelper.hasVerifiedAccount()) {
                syncDropbits();
            }
            updateWallet();
        } catch (DaoException e) {
            e.printStackTrace();
        }

        localBroadCastUtil.sendGlobalBroadcast(WalletSyncCompletedReceiver.class,
                Intents.ACTION_WALLET_SYNC_COMPLETE);
    }

    private void syncTransactions() {
        accountDeverificationServiceRunner.run();
        walletRegistraionRunner.run();
        currentBTCStateRunner.run();
        syncRunnable.run();
        transactionConfirmationUpdateRunner.run();
    }

    private void syncDropbits() {
        if (!walletHelper.hasVerifiedAccount()) return;
        syncIncomingInvitesRunner.run();
        fulfillSentInvitesRunner.run();
        receivedInvitesStatusRunner.run();
        negativeBalanceRunner.run();
        remoteAddressCache.cacheAddresses();
    }

    private void updateWallet() {
        failedBroadcastCleaner.run();
        cnWalletManager.updateBalances();
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/service/runner/FullSyncWalletRunner.kt
package com.coinninja.coinkeeper.service.runner

import app.coinninja.cn.thunderdome.repository.ThunderDomeRepository
import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.cn.account.RemoteAddressCache
import com.coinninja.coinkeeper.cn.dropbit.DropBitMeServiceManager
import com.coinninja.coinkeeper.cn.service.runner.AccountDeverificationServiceRunner
import com.coinninja.coinkeeper.cn.transaction.LightningInviteLinker
import com.coinninja.coinkeeper.cn.transaction.LightningWithdrawalLinker
import com.coinninja.coinkeeper.cn.wallet.CNWalletManager
import com.coinninja.coinkeeper.model.helpers.DropbitAccountHelper
import com.coinninja.coinkeeper.receiver.WalletSyncCompletedReceiver
import com.coinninja.coinkeeper.util.DropbitIntents
import com.coinninja.coinkeeper.util.android.LocalBroadCastUtil

@Mockable
class FullSyncWalletRunner constructor(internal val cnWalletManager: CNWalletManager,
                                       internal val accountDeverificationServiceRunner: AccountDeverificationServiceRunner,
                                       internal val walletRegistrationRunner: WalletRegistrationRunner,
                                       internal val currentBTCStateRunner: CurrentBTCStateRunner,
                                       internal val syncRunnable: SyncRunnable,
                                       internal val transactionConfirmationUpdateRunner: TransactionConfirmationUpdateRunner,
                                       internal val failedBroadcastCleaner: FailedBroadcastCleaner,
                                       internal val syncIncomingInvitesRunner: SyncIncomingInvitesRunner,
                                       internal val fulfillSentInvitesRunner: FulfillSentInvitesRunner,
                                       internal val receivedInvitesStatusRunner: ReceivedInvitesStatusRunner,
                                       internal val negativeBalanceRunner: NegativeBalanceRunner,
                                       internal val dropbitAccountHelper: DropbitAccountHelper,
                                       internal val localBroadCastUtil: LocalBroadCastUtil,
                                       internal val remoteAddressCache: RemoteAddressCache,
                                       internal val dropBitMeServiceManager: DropBitMeServiceManager,
                                       internal val thunderDomeRepository: ThunderDomeRepository,
                                       internal val lightningWithdrawlLinker: LightningWithdrawalLinker,
                                       internal val lightningInviteLinker: LightningInviteLinker
) : Runnable {

    override fun run() {
        if (!cnWalletManager.hasWallet)
            return

        try {
            accountDeverificationServiceRunner.run()
            walletRegistrationRunner.run()
            currentBTCStateRunner.run()
            dropBitMeServiceManager.syncIdentities()

            syncDropbits()

            syncRunnable.run()

            if (!cnWalletManager.isSegwitUpgradeRequired) {
                thunderDomeRepository.sync()
            }
            lightningInviteLinker.linkInvitesToInvoices()
            lightningWithdrawlLinker.linkWithdraws()
            transactionConfirmationUpdateRunner.run()
            failedBroadcastCleaner.run()
            cnWalletManager.updateBalances()
        } catch (e: Exception) {
            e.printStackTrace()
        }

        localBroadCastUtil.sendGlobalBroadcast(WalletSyncCompletedReceiver::class.java,
                DropbitIntents.ACTION_WALLET_SYNC_COMPLETE)
    }

    private fun syncDropbits() {
        if (!dropbitAccountHelper.hasVerifiedAccount) return

        syncIncomingInvitesRunner.run()
        receivedInvitesStatusRunner.run()
        negativeBalanceRunner.run()
        fulfillSentInvitesRunner.run()
        remoteAddressCache.cacheAddresses()
    }
}
